File: src/bin/day07.rs
#![feature(str_split_once)]

extern crate nom;
use std::collections::HashMap;
use std::fs;

struct BagCount<'a> {
    name: &'a str,
    count: i32,
}

fn parse_line(input: &str) -> (&str, Vec<BagCount>) {
    let split0 = input
        .trim_end_matches('.')
        .split(" bags contain ")
        .collect::<Vec<_>>();
    let this_bag = split0[0];
    if split0[1] == "no other bags" {
        return (this_bag, vec![]);
    }
    let contains = split0[1]
        .split(", ")
        .map(|bag| {
            let parts = bag.split_once(" ").unwrap();
            let name = parts.1.split(" bag").next().unwrap();
            BagCount {
                count: parts.0.parse::<i32>().unwrap(),
                name,
            }
        })
        .collect();
    (this_bag, contains)
}

fn can_hold_my_bag<'a>(
    some_bag: &'a str,
    my_bag: &'a str,
    memory: &mut HashMap<&'a str, bool>,
    all_bags: &HashMap<&'a str, Vec<BagCount<'a>>>,
) -> bool {
    if *memory.get(some_bag).unwrap_or(&false) {
        return true;
    }
    let contents = all_bags.get(some_bag).unwrap();
    for bag_count in contents {
        if bag_count.name == my_bag {
            memory.insert(some_bag, true);
            return true;
        }
        if can_hold_my_bag(bag_count.name, my_bag, memory, all_bags) {
            memory.insert(some_bag, true);
            return true;
        }
    }
    memory.insert(some_bag, false);
    false
}

fn bag_counter<'a>(
    some_bag: &'a str,
    memory: &mut HashMap<&'a str, i32>,
    all_bags: &HashMap<&'a str, Vec<BagCount<'a>>>,
) -> i32 {
    if let Some(prev) = memory.get(some_bag) {
        return *prev;
    }
    let counter = 1 + all_bags
        .get(some_bag)
        .unwrap()
        .iter()
        .map(|bag_count| bag_count.count * bag_counter(bag_count.name, memory, all_bags))
        .sum::<i32>();
    memory.insert(some_bag, counter);
    counter
}

fn main() {
    let contents = fs::read_to_string("inputs/day07.txt").expect("Something went wrong reading the file");

    let bags: HashMap<&str,Vec<BagCount>> = contents.lines().map(|ln| parse_line(ln)).collect();

    // for bag in &bags {
    //     println!("{} has bags:", bag.0);
    //     for b in bag.1 {
    //         println!(" - {} {}", b.count, b.name);
    //     }
    // }
    let mut can_contain_shiny = HashMap::new();
    for name in bags.keys() {
        can_hold_my_bag(name, "shiny gold", &mut can_contain_shiny, &bags);
    }
    println!(
        "{} bags can contain my shiny gold bag",
        can_contain_shiny.values().filter(|b| **b).count()
    );
    let mut bag_contents = HashMap::new();
    println!(
        "{} bags inside shiny gold bag",
        bag_counter("shiny gold", &mut bag_contents, &bags) - 1
    );
}

File: julia/day07.jl
lines = open("../inputs/day07.txt") do f
    readlines(f)
end

struct BagCount
    name::String
    count::Int64
end

function parse_line(line::String)::Tuple{String,Vector{BagCount}}
    ln = rstrip(line, '.')
    parts = split(ln, " bags contain ")
    @assert (length(parts) == 2)
    name = parts[1]
    contents = parts[2]
    if contents == "no other bags"
        return (name, [])
    end

    return (name, [
        begin
        parts = split(bag, " ", limit=2)
        name = split(parts[2], " bag")[1]
        count = parse(Int64, parts[1])
        BagCount(name, count)
    end
        for bag in split(contents, ", ")
    ])
end

function canHoldMyBag(someBag, myBag, memory, bagDict)::Bool
    prev = get(memory, someBag, nothing)
    if prev !== nothing
        return prev
    end
    for bag in bagDict[someBag]
        if bag.name == myBag || canHoldMyBag(bag.name, myBag, memory, bagDict)
            memory[someBag] = true
            return true
        end
    end
    memory[someBag] = false
    return false
end

function bagCounter(someBag, memory, bagDict)::Int64
    prev = get(memory, someBag, nothing)
    if prev !== nothing
        return prev
    end
    return 1 + sum([ bag.count * bagCounter(bag.name, memory, bagDict)
                     for bag in bagDict[someBag] ])
end

function part1(bagDict, myBag)
    memory = Dict{String,Bool}()
    return length([ ()
                    for name in keys(bagDict)
                    if canHoldMyBag(name, myBag, memory, bagDict)])
end

function part2(bagDict, myBag)
    memory = Dict{String,Int64}()
    return bagCounter(myBag, memory, bagDict) - 1
    
end

myBag = "shiny gold"
bags = parse_line.(lines)
bagDict = Dict{String,Vector{BagCount}}(
    name => contents
    for (name, contents) in bags)

println("Part 1: ", part1(bagDict, myBag))
println("Part 2: ", part2(bagDict, myBag))